Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp

```diff
@@ -902,98 +975,71 @@ namespace AzToolsFramework
         }
     }
 
-    void DocumentPropertyEditor::SetSavedExpanderStateForRow(DPERowWidget* row, DocumentPropertyEditor::ExpanderState expanderState)
+    void DocumentPropertyEditor::SetSavedStateKey(AZ::u32 key, AZStd::string propertyEditorName)
     {
-        // Get the index of each dom child going up the chain. We can then reverse this
-        // and use these indices to mark a path to expanded/collapsed nodes
-        AZStd::vector<size_t> reversePath = GetPathToRoot(row);
+        // We need to append some alphabetical characters to the key or it will be treated as a very large json array index
+        AZStd::string_view keyStr = AZStd::string::format("uuid%s", AZStd::to_string(key).c_str());
+        m_dpeSettings = AZStd::make_unique<DocumentPropertyEditorSettings>(keyStr, propertyEditorName);
 
-        if (!reversePath.empty())
+        if (m_dpeSettings && m_dpeSettings->WereSettingsLoaded())
         {
-            // create new pathNodes when necessary implicitly by indexing into each map,
-            // then set the expander state on the final node
-            auto reverseIter = reversePath.rbegin();
-            auto* currPathNode = &m_expanderPaths[*(reverseIter++)];
+            m_dpeSettings->SetCleanExpanderStateCallback(
+                [this](DocumentPropertyEditorSettings::ExpanderStateMap& storedStates)
+                {
+                    bool wereChangesMade = false;
+                    auto rootValue = m_adapter->GetContents();
+                    auto iter = storedStates.begin();
+                    while (iter != storedStates.end())
+                    {
+                        if (auto value = rootValue.FindChild(AZ::Dom::Path(iter->first)); !value)
+                        {
+                            iter = storedStates.erase(iter);
+                            wereChangesMade = true;
+                        }
+                        else
+                        {
+                            ++iter;
+                        }
+                    }
```

Comment:
This is the erase idiom, which should be done using the [AZStd::erase_if](https://en.cppreference.com/w/cpp/container/unordered_map/erase_if) function

```suggestion
                    auto EraseStates = [&changesWereMade, &rootValue](const AZStd::pair<AZStd::string, bool>& statePair)
                    {
                        if (auto value = rootValue.FindChild(AZ::Dom::Path(statePair.first)); !value)
                        {
                            wereChangesMade = true;
                            return true;
                        }
                        else
                        {
                            return false;
                        }
                    };
                    
                    AZStd::erase_if(storedStates, EraseStates);
```

Reply:
Followed suggestion with some adjustments. Also fixed the param that I dropped the const& on.